README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![Travis-CI Build
Status](https://travis-ci.org/leslie-huang/stylest.svg?branch=master)](https://travis-ci.org/leslie-huang/stylest)
[![Build
status](https://ci.appveyor.com/api/projects/status/384insyingv6jcg2?svg=true)](https://ci.appveyor.com/project/leslie-huang/stylest)

# stylest version 0.2.0

`stylest` estimates speaker style distinctiveness.

## Installation

Install `stylest` from CRAN by executing:

``` r
install.packages("stylest")
```

The dev version of `stylest` on GitHub may have additional features (and
bugs) and is not guaranteed to be stable. Power users may install it
with:

``` r
# install.packages("devtools")
# devtools::install_github("leslie-huang/stylest")
```

See the
<a href="https://leslie-huang.github.io/stylest/articles/stylest-vignette.html">vignette</a>
for examples and usage.

## Authors

<a href="http://leslie-huang.github.io">Leslie Huang (maintainer)</a>

<a href="https://github.com/patperry">Patrick O. Perry</a>

<a href="https://github.com/ArthurSpirling/">Arthur Spirling</a>

## Bug reports

Please submit any bugs, error reports, etc. on
<a href="https://github.com/leslie-huang/stylest/issues">GitHub</a>.


NEWS.md
# stylest 0.2.0

This update adds the option for custom term weights to be added to the main `stylest_fit()` function.

# stylest 0.1.0

This is the first CRAN release of the stylest package for estimating speaker style distinctiveness.


tests/testthat.R
library(testthat)
library(stylest)

test_check("stylest")


tests/testthat/test-stylest_fit_speaker.R
library(stylest)

test_that("invalid smooth raises error", {
  expect_error(stylest_select_vocab(novels_excerpts, novels_excerpts$author, smooth = -1),
               "smooth value must be greater than or equal to 0")
}
)

tests/testthat/test-stylest_select_vocab.R
library(stylest)

test_that("invalid nfolds raises error", {
  expect_error(stylest_select_vocab(novels_excerpts, novels_excerpts$author, nfold = 0),
               "nfolds must be at least 1")
  expect_error(stylest_select_vocab(novels_excerpts, novels_excerpts$author, nfold = 0.5),
               "nfold must be an integer value")
  }
)

test_that("invalid smooth raises error", {
  expect_error(stylest_select_vocab(novels_excerpts, novels_excerpts$author, smooth = -1),
               "smooth value must be greater than or equal to 0")
}
)

test_that("invalid cutoff pct raises error", {
  expect_error(stylest_select_vocab(novels_excerpts, novels_excerpts$author, cutoff_pcts = c(-50, 100)),
               "cutoff percent must be value between 0 and 100")
}
)


R/generics.R
#' Summarise Predictions
#'
#' This summary function returns a data frame with documents as
#' rownames and for each document, either its the most likely author 
#' or the probability that it was written by each possible author.
#' 
#' If \code{type} is "\code{predicted}",
#' the columns are "\code{author}", a character vector of predicted author
#' names, and "\code{prob}", the posterior probability of this author
#' assignment under the model.
#' If \code{type} is "\code{logprobs}" columns are named by
#' possible author and each element is
#' the log posterior probability assigned to each document and author,
#' relative to the other possible authors.
#' If \code{type} is "\code{probs}", each element is the posterior
#' probability that a document is assigned to each author relative to
#' the others.  This is constructed from the log posterior
#' using a logistic transformation across rows, a.k.a. 'softmax'.
#' 
#' @param object Predictions from a stylest_model
#'               (a \code{\link{stylest_predict}} object)
#' @param type Type of prediction "\code{predicted}" (the default),
#'             "\code{logprobs}" or "\code{probs}".  See Details.
#' @param ... Other arguments (ignored)
#'
#' @return a summary data frame of predictions
#' @export
#' @examples
#' data(novels_excerpts)
#' speaker_mod <- stylest_fit(novels_excerpts$text, novels_excerpts$author)
#' preds <- stylest_predict(speaker_mod, "This is an example text, who wrote it?")
#' summary(preds)
summary.stylest_predict <- function(object,
                                    type = c("predicted", "probs", "logprobs"),
                                    ...){
  type <- match.arg(type)
  m <- as.data.frame(as.matrix(object$log_probs))
  if (type != "logprobs"){
    m <- exp(as.matrix(object$log_probs))
    m <- as.data.frame(m / rowSums(m))
    if (type == "predicted")
      m <- data.frame(author = as.character(object$predicted),
                      prob = apply(m, 1, max),
                      stringsAsFactors = FALSE)
  }
  m
}

#' Predict the Authors of Documents
#'
#' This function returns a data frame with documents as
#' rownames and author predictions as columns.
#' 
#' If \code{type} is "\code{predicted}" the columns are
#' "\code{author}", a character vector of predicted author names, and
#' "\code{prob}", the posterior probability of this author assignment under the model.
#' If \code{type} is "\code{logprobs}" there are as many columns as possible
#' authors and each element of the data frame is
#' the log posterior probability the model assigns to each author and
#' document.  If \code{type} is "\code{probs}", each element is the posterior
#' probability that a document is assigned to an author relative to all the
#' others.  This is constructed from the log posterior by inverse logistic
#' transformation, a.k.a. 'softmax'.
#'
#' @param object A fitted stylest_model
#' @param newdata A character vector of texts or a \code{\link{corpus_frame}}
#' @param type Type of prediction "\code{predicted}" (the default),
#'             "\code{logprobs}" or "\code{probs}". See below for details
#' @param ... Other arguments (ignored)
#' @return a data frame of author predictions
#' @export
#' @examples
#' data(novels_excerpts)
#' speaker_mod <- stylest_fit(novels_excerpts$text, novels_excerpts$author)
#' predict(speaker_mod, "This is an example text, who wrote it?", type = "probs")
predict.stylest_model <- function(object, newdata,
                                  type = c("predicted", "probs", "logprobs"),
                                  ...){
  type <- match.arg(type)
  if ("corpus_frame" == class(newdata)){
    newdata <- newdata$text
    names(newdata) <- newdata$title
  }
  preds <- stylest_predict(object, newdata)
  m <- as.data.frame(as.matrix(preds$log_probs))
  if (type != "logprobs"){
    m <- exp(as.matrix(preds$log_probs))
    m <- as.data.frame(m / rowSums(m))
    if (type == "predicted")
      m <- data.frame(author = as.character(preds$predicted),
                      prob = apply(m, 1, max),
                      stringsAsFactors = FALSE)
  }
  m
}


R/stylest_fit.R
#' Fit speaker_model to a corpus
#'
#' The main function in \code{stylest}, \code{stylest_fit} fits a
#' model using a corpus of texts labeled by speaker.
#' 
#' The user may specify only one of \code{terms} or \code{cutoff}.
#' If neither is specified, all terms will be used.
#'
#' @export
#' 
#' @param x Text vector. May be a \code{corpus_frame} object
#' @param speaker Vector of speaker labels. Should be the same length as
#'   \code{x}
#' @param terms If not \code{NULL}, terms to be used in the model. If
#'   \code{NULL}, use all terms
#' @param filter If not \code{NULL}, a text filter to specify the tokenization.
#' See \code{corpus} for more information about specifying \code{filter}
#' @param smooth Numeric value used smooth term frequencies instead of the
#'   default of 0.5
#' @param term_weights Dataframe of distances (or any weights) per 
#' word in the vocab. This dataframe should have one column $word and 
#' a second column $weight_var containing the weight for the word.
#' See the vignette for details.
#' @param fill_weight numeric value to fill in as weight for any term
#' which does not have a weight specified in \code{term_weights}, default=\code{1.0}
#' @param weight_varname Name of the column in term_weights containing the weights, default=\code{"mean_distance"}
#' @return A S3 \code{stylest_model} object containing:
#' \code{speakers} Vector of unique speakers,
#' \code{filter} text_filter used,
#' \code{terms} terms used in fitting the model,
#' \code{ntoken} Vector of number of tokens per speaker,
#' \code{smooth} Smoothing value,
#' \code{weights} If not NULL, a named matrix of weights for each term in the vocab,
#' \code{rate} Matrix of speaker rates for each term in vocabulary
#' 
#' @examples
#' data(novels_excerpts)
#' speaker_mod <- stylest_fit(novels_excerpts$text, novels_excerpts$author)
#' 
stylest_fit <- function(x, speaker, terms = NULL, filter = NULL, smooth = 0.5, 
                        term_weights = NULL, weight_varname = "mean_distance",
                        fill_weight=1.0)
{
  
  if (smooth <= 0) {
    stop("smooth value must be greater than or equal to 0")
  }
  
  {
    # coerce arguments to their expected types
    x <- corpus::as_corpus_text(x, filter)
    speaker <- as.factor(speaker)

    if (is.null(terms)) {
        # default to fitting a model with all terms
        terms <- corpus::text_types(x, collapse = TRUE)
    } else {
        terms <- as.character(terms)
    }

    smooth <- as.numeric(smooth)[[1]]

    # fit the model
    model <- fit_term_usage(x, speaker, terms, smooth, term_weights, 
                            weight_varname, fill_weight)
    cl <- "stylest_model_term"

    # package everything in an object
    result <- structure(c(list(speakers = levels(speaker),
                              filter = corpus::text_filter(x)),
                          model), 
                       class = c(cl, "stylest_model"))
    
    result
  }
}

#' Computes speakers' term usage rates
#'
#' @param x Text vector. May be a \code{corpus_frame} object
#' @param speaker Vector of speaker labels. Should be the same length as
#'   \code{x}
#' @param terms Vocabulary for document term matrix
#' @param smooth Numeric value used smooth term frequencies
#' @param term_weights Dataframe of distances (or any weights) per 
#' word in the vocab. This dataframe should have one column $word and 
#' a second column $weight_var containing the weight for the word
#' @param weight_varname Name of the column in term_weights containing the weights
#' @return named list of terms, vector of num tokens uttered by each speaker,
#'   smoothing value, term weights (NULL if no weights), and (smoothed) term usage rate matrix
#' @param fill_weight numeric value to fill in as weight for any term
#' which does not have a weight specified in \code{term_weights}
#'   
fit_term_usage <- function(x, speaker, terms, smooth, term_weights, weight_varname, fill_weight)
{
    # get a term matrix for the selected terms and selected speaker
    selected_dtm <- corpus::term_matrix(x, select = terms, group = speaker)

    # get the number of tokens uttered by each speaker
    ntok <- Matrix::rowSums(selected_dtm)

    # get the number of unique types in the corpus
    ntype <- ncol(selected_dtm)

    # compute the (smoothed) usage rates
    rate <- (as.matrix(selected_dtm) + smooth) / (ntok + smooth * ntype)

    if (is.null(term_weights)) {
      weights <- NULL
    }
    else {
      # construct vector of multipliers for words in same order as rate
      weights <- c()
      for (word in colnames(rate)) {
        if (word %in% term_weights$word) {
          weights <- c(weights, 
                       term_weights[term_weights$word == word, 
                                           weight_varname])
        }
        else {
          weights <- c(weights, fill_weight)
        }
      }
      names(weights) <- colnames(rate)
      
      # multiply each word's rate by its weight
      rate <- t(t(rate) * weights)
    }
    
    
    list(terms = terms, ntoken = ntok, smooth = smooth, weights = weights, rate = rate)
}

#'
#' Custom print method for stylest_model
#' 
#' @export
#' 
#' @param x `stylest_model` object
#' @param ... Additional arguments
#' @return Prints summary information about the `stylest_model` object
#' 
#' @examples 
#' 
#' data(novels_excerpts)
#' speaker_mod <- stylest_fit(novels_excerpts$text, novels_excerpts$author)
#' print(speaker_mod)
#'  
print.stylest_model <- function(x, ...) {
  cat("A S3 stylest_model object containing: ")
  cat(length(x$speakers), "unique authors and", length(x$terms), "unique terms.")
  
}

R/stylest_predict.R
#' Predict the most likely speaker of a text
#'
#' Use a fitted stylest_model to predict the most likely speaker of a text.
#' This function may be used on in-sample or out-of-sample texts.
#'
#' @export
#'
#' @param model \code{stylest_model} object
#' @param text Text vector. May be a \code{corpus_frame} object
#' @param prior Prior probability, defaults to \code{NULL}
#' @return \code{stylest_predict} object containing:
#' \code{model} the fitted \code{stylest_model} object used in prediction,
#' \code{predicted} the predicted speaker,
#' \code{log_probs} matrix of log probabilities,
#' \code{log_prior} matrix of log prior probabilities
#' 
#' @examples 
#' data(novels_excerpts)
#' speaker_mod <- stylest_fit(novels_excerpts$text, novels_excerpts$author)
#' stylest_predict(speaker_mod, "This is an example text, who wrote it?")
#'   
stylest_predict <- function(model, text, prior = NULL)
{
    stopifnot(inherits(model, "stylest_model"))

    nspeaker <- length(model$speakers)

    text <- corpus::as_corpus_text(text, model$filter)

    # use an equal prior if left unspecified
    if (is.null(prior)) {
        log_prior <- rep(-log(nspeaker), nspeaker) # log(1/nspeaker)
    } else {
        log_prior <- log(prior)
    }
    names(log_prior) <- model$speakers

    terms <- model$terms
    x <- corpus::term_matrix(text, select = model$terms)
    ntok <- rowSums(x)

    # for each speech, evaluate the Poisson log-likelihoods for
    # the speakers; don't worry about the normalizing constant since
    # it is the same for all speakers
    eta <- log(model$rate)
    loglik <- x %*% t(eta) - ntok %*% t(rowSums(model$rate))
    rownames(loglik) <- rownames(x)

    # for each speech, compute the posterior probabilities for the speakers
    log_weights <- loglik
    if (!is.null(prior)) {
        log_weights <- sweep(log_weights, 2, log_prior, "+")
    }

    # guard against overflow
    log_weights <- log_weights - apply(log_weights, 1, max)

    const <- log(Matrix::rowSums(exp(log_weights)))
    log_probs <- log_weights - const
    rownames(log_probs) <- rownames(x)

    predicted <- model$speakers[apply(log_probs, 1, which.max)]
    predicted <- factor(predicted, levels = model$speakers)

    res <- list(model = model, predicted = predicted,
                log_probs = log_probs, log_prior = log_prior)
    class(res) <- "stylest_predict"
    res
}


R/stylest_speaker_odds.R
#' Pairwise prediction of the most likely speaker of texts
#'
#' Computes the mean log odds of the most likely speaker of each
#' text over pairs of the speaker of a text and every other speaker in the
#' \code{stylest_model}.
#'
#' @export
#'
#' @param model \code{stylest_model} object
#' @param text Text vector. May be a \code{corpus_frame} object
#' @param speaker Vector of speaker labels. Should be the same length as
#'   \code{x}
#' @param prior Prior probability of speakers. Uses equal prior if \code{NULL}
#' @return A S3 \code{stylest_odds} object containing: a
#'   \code{stylest_model} object; vector of mean log odds that each actual
#'   speaker (compared with other speakers in the corpus) spoke their
#'   corresponding texts in the corpus; vector of SEs of the log odds
#'
#' @examples
#' data(novels_excerpts)
#' speaker_mod <- stylest_fit(novels_excerpts$text, novels_excerpts$author)
#' stylest_odds(speaker_mod, novels_excerpts$text, novels_excerpts$author)
#'   
stylest_odds <- function(model, text, speaker, prior = NULL)
{
    stopifnot(inherits(model, "stylest_model"))
  
    nspeaker <- length(model$speakers)

    text <- corpus::as_corpus_text(text, model$filter)
    baseline <- match(as.character(speaker), model$speakers)

    # use an equal prior if left unspecified
    if (is.null(prior)) {
        log_prior <- rep(-log(nspeaker), nspeaker) # log(1/nspeaker)
    } else {
        log_prior <- log(prior)
    }
    names(log_prior) <- model$speakers

    terms <- model$terms
    x <- corpus::term_matrix(text, select = model$terms)
    ntok <- Matrix::rowSums(x)

    # for each speech, evaluate the Poisson log-likelihoods for
    # the speakers; don't worry about the normalizing constant since
    # it is the same for all speakers
    eta <- log(model$rate)
    loglik <- x %*% t(eta) - ntok %*% t(Matrix::rowSums(model$rate))
    rownames(loglik) <- rownames(x)

    # for each speech, compute the posterior probabilities for the speakers
    log_weights <- loglik
    if (!is.null(prior)) {
        log_weights <- sweep(log_weights, 2, log_prior, "+")
    }

    # NOTE: normalize by number of tokens to make speeches of different
    # lengths comparable
    log_weights <- log_weights / ifelse(ntok == 0, 1, ntok)

    # compute ratio of baseline to other (zeros are actual spkr)
    log_odds <- log_weights[cbind(seq_along(baseline), baseline)] - log_weights

    # average globally
    nspk <- ncol(log_odds)
    log_odds_avg <- Matrix::rowSums(log_odds) / nspk
    log_odds_se  <- apply(log_odds, 1, stats::sd) / sqrt(nspk)

    # put NA for speeches of length 0
    log_odds_avg[ntok == 0] <- NA
    log_odds_se[ntok == 0] <- NA

    res <- list(model = model, log_odds_avg = log_odds_avg,
                log_odds_se = log_odds_se)
    class(res) <- "stylest_odds"
    res
}


R/stylest_term_influence.R
#' Compute the influence of terms
#' 
#' @export
#'
#' @param model \code{stylest_model} object
#' @param text Text vector. May be a \code{corpus_frame} object
#' @param speaker Vector of speaker labels. Should be the same length as
#'   \code{x}
#' @return \code{data.frame} with columns representing terms, their mean influence,
#' and their maximum influence
#'   
#' @examples 
#' data(novels_excerpts)
#' speaker_mod <- stylest_fit(novels_excerpts$text, novels_excerpts$author)
#' stylest_term_influence(speaker_mod, novels_excerpts$text, novels_excerpts$author)
stylest_term_influence <- function(model, text, speaker)
{
  if (!inherits(model, "stylest_model_term"))
    stop("unsupported stylest model type")
  
    eta <- log(model$rate)
    text <- corpus::as_corpus_text(text, model$filter)

    x <- corpus::term_matrix(text, select = model$terms)
    ntok <- Matrix::rowSums(x)

    fbar <- matrix(NA, length(model$speakers), length(model$terms))
    rownames(fbar) <- model$speakers
    colnames(fbar) <- model$terms

    for (i in seq_len(nrow(fbar))) {
        t <- model$speakers[[i]]
        x_t <- x[speaker == t, , drop = FALSE]
        fbar[i, ] <- Matrix::colMeans(x_t) #spkr's mean term frequency rate (averaged over their docs)
    }

    etabar <- Matrix::colMeans(eta)
    eta_centered <- eta - matrix(1, nrow(eta), 1) %*% etabar

    d <- fbar * eta_centered

    term <- model$terms
    infl_avg <- apply(abs(d), 2, mean)
    infl_max <- apply(abs(d), 2, max)

    data.frame(term, infl_avg, infl_max, row.names = NULL)
}


R/stylest_select_vocab.R
#' Select vocabulary using cross-validated out-of-sample prediction
#'
#' Selects optimal vocabulary quantile(s) for model fitting using performance on
#' predicting out-of-sampletexts.
#'
#' @export
#'
#' @param x Corpus as text vector. May be a \code{corpus_frame} object
#' @param speaker Vector of speaker labels. Should be the same length as
#'   \code{x}
#' @param filter if not \code{NULL}, a \code{corpus} text_filter
#' @param smooth value for smoothing. Defaults to 0.5
#' @param nfold Number of folds for cross-validation. Defaults to 5
#' @param cutoff_pcts Vector of cutoff percentages to test. Defaults to
#'   \code{c(50, 60, 70, 80, 90, 99)}
#' @return List of: best cutoff percent with the best speaker classification
#'   rate; cutoff percentages that were tested; matrix of the mean percentage of
#'   incorrectly identified speakers for each cutoff percent and fold; and the
#'   number of folds for cross-validation
#'
#' @examples
#' \dontrun{
#' data(novels_excerpts)
#' stylest_select_vocab(novels_excerpts$text, novels_excerpts$author, cutoff_pcts = c(50, 90))
#' }
#'   
stylest_select_vocab <- function(x, speaker, filter = NULL, smooth = 0.5, nfold = 5,
                             cutoff_pcts = c(50, 60, 70, 80, 90, 99)) {

  if (as.integer(nfold) != nfold) {
    stop("nfold must be an integer value")
  }
  
  if (as.integer(nfold) < 1 ) {
    stop("nfolds must be at least 1")
  }

  if (smooth <= 0) {
    stop("smooth value must be greater than or equal to 0")
  }
  
  for (c in cutoff_pcts) {
    if (c < 0 | c >= 100) {
      stop("cutoff percent must be value between 0 and 100")
    }
  }
  
  {
  # coerce arguments to their expected types
  x <- corpus::as_corpus_text(x, filter)
  speaker <- as.factor(speaker)
  
  ntot <- length(x)
  test_fold <- sample(rep(1:nfold, ceiling(ntot / nfold)), ntot)
  
  miss_pct <- matrix(NA, nfold, length(cutoff_pcts))
  
  for (fold in 1:nfold)  {
    # set up test and training sets
    test_set <- (test_fold == fold)
    train_set <- !test_set
    test <- x[test_set]
    train <- x[train_set]
    
    speech_stats <- corpus::term_stats(train)
    
    for (i in seq_along(cutoff_pcts)) {
      # select subset of vocab above cutoff percent
      cutoff_pct <- cutoff_pcts[[i]]
      cutoff <- cutoff_pct / 100
      terms <- subset(speech_stats, speech_stats$support >= quantile(speech_stats$support, cutoff))$term
      
      # fit model on training data 
      fit <- stylest_fit(train, speaker[train_set], terms, smooth = smooth) 
      # predict speaker for test data
      pred <- stylest_predict(fit, test) 
      # mean num incorrectly predicted speakers on test data per fold
      miss_pct[fold, i] <- 100 * mean(pred$predicted != speaker[test_set]) 
    }
  }
  
  # find the cutoff percent with the lowest miss percentage
  avg <- apply(miss_pct, 2, mean)
  i <- which.min(avg)
  
  res <- list(cutoff_pct_best = cutoff_pcts[i],
       cutoff_pcts = cutoff_pcts,
       miss_pct = miss_pct,
       nfold = nfold)
  class(res) <- "stylest_select_vocab"
  
  res
  }
}


#' Use vocab cutoff to select terms for fitting the model
#' 
#' The same text, speaker, and filter should be used in this model
#' as in \code{fit_speaker} to select the terms for the latter function.
#' 
#' @export
#' 
#' @param x Corpus as text vector. May be a \code{corpus_frame} object
#' @param speaker Vector of speaker labels. Should be the same length as
#'   \code{x}
#' @param vocab_cutoff Quantile cutoff for the vocabulary in (0, 100]
#' @param filter if not \code{NULL}, a corpus filter
#' @return list of terms
#' @examples 
#' data(novels_excerpts)
#' stylest_terms(novels_excerpts$text, novels_excerpts$author, vocab_cutoff = 50)
#'  
stylest_terms <- function(x, speaker, vocab_cutoff, filter = NULL) {
  
  if (vocab_cutoff < 0 | vocab_cutoff > 100) {
    stop("vocab cutoff percent must be between 0 and 100")
  }
  
  {
    speeches <- corpus::corpus_frame(speaker = speaker,
                             text = x,
                             filter = filter)
    
    t_stats <- corpus::term_stats(speeches)
    
    cutoff_pct <- vocab_cutoff / 100
    
    terms <- subset(t_stats, t_stats$support >= quantile(t_stats$support, cutoff_pct))$term
    
    terms
  }
}